src/dev/pi4ioe5v6408.rs

```diff
@@ -0,0 +1,271 @@
+            addr,
+            no_initialize,
+        )?)))
+    }
+
+    pub fn split<'a>(&'a mut self) -> Parts<'a, I2C, M> {
+        Parts {
+            io0: crate::Pin::new(0, &self.0),
+            io1: crate::Pin::new(1, &self.0),
+            io2: crate::Pin::new(2, &self.0),
+            io3: crate::Pin::new(3, &self.0),
+            io4: crate::Pin::new(4, &self.0),
+            io5: crate::Pin::new(5, &self.0),
+            io6: crate::Pin::new(6, &self.0),
+            io7: crate::Pin::new(7, &self.0),
+        }
+    }
+}
+
+pub struct Parts<'a, I2C, M = shared_bus::NullMutex<Driver<I2C>>>
+where
+    I2C: crate::I2cBus,
+    M: shared_bus::BusMutex<Bus = Driver<I2C>>,
+{
+    pub io0: crate::Pin<'a, crate::mode::Input, M>,
+    pub io1: crate::Pin<'a, crate::mode::Input, M>,
+    pub io2: crate::Pin<'a, crate::mode::Input, M>,
+    pub io3: crate::Pin<'a, crate::mode::Input, M>,
+    pub io4: crate::Pin<'a, crate::mode::Input, M>,
+    pub io5: crate::Pin<'a, crate::mode::Input, M>,
+    pub io6: crate::Pin<'a, crate::mode::Input, M>,
+    pub io7: crate::Pin<'a, crate::mode::Input, M>,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Regs {
+    DeviceIdControl = 0x01,
+    IODirection = 0x03,
+    OutputPort = 0x05,
+    OutputHighImpedance = 0x07,
+    InputDefaultState = 0x09,
+    PullUpPullDownEnable = 0x0b,
+    PullUpPullDownSelection = 0x0d,
+    InputStatusRegister = 0x0f,
+    InterruptMaskRegister = 0x11,
+    InterruptStatusRegister = 0x13,
+}
+
+impl From<Regs> for u8 {
+    fn from(r: Regs) -> u8 {
+        r as u8
+    }
+}
+
+pub struct Driver<I2C> {
+    i2c: I2C,
+    addr: u8,
+    out: u8,
+}
+
+impl<I2C: crate::I2cBus> Driver<I2C> {
+    pub fn new(mut i2c: I2C, addr: bool, no_initialize: bool) -> Result<Self, I2C::BusError> {
+        let addr = if addr { 0x44 } else { 0x43 };
+
+        let device_id = i2c.read_reg(addr, Regs::DeviceIdControl)?; // Reset the "(Power on) Reset Interrupt" bit (and validate the device ID)
+        assert_eq!(
+            device_id & 0xFC, // Only check Manufacturer ID (0b101) and Firmware Revision (0b000)
+            0xA0,
+            "Unexpected Device ID for the PI4IOE5V6408: 0x{:02x}",
+            device_id
+        );
+
+        // The Reset values are the following:
+
+        // i2c.write_reg(addr, Regs::IODirection, 0)?; // All pins as inputs
+        // i2c.write_reg(addr, Regs::OutputPort, 0)?; // Set all outputs to low
+        // i2c.write_reg(addr, Regs::OutputHighImpedance, 0xff)?; // Set high impedance mode on all outputs
+        // i2c.write_reg(addr, Regs::InputDefaultState, 0)?; // The default state of all inputs is 0
+        // i2c.write_reg(addr, Regs::PullUpPullDownEnable, 0xff)?; // Pull-Up/Pull-Down enabled on all inputs
+        // i2c.write_reg(addr, Regs::PullUpPullDownSelection, 0)?; // Pull-Downs on all inputs
+        // i2c.write_reg(addr, Regs::InterruptMaskRegister, 0)?; // Interrupts enabled on all inputs
```

Comment:
So this gets me thinking whether we should actually be resetting all registers by default for all of our supported port expanders.  I'd say it is generally good practice to ensure the device is entirely in the state you expect it to be in.

What do you think about this?  If you agree, I'd open an issue for the topic.

Reply:
Yes, I think it makes sense to do that (in a separate issue/pr)